client/lib/route/path.ts

```diff
@@ -41,8 +41,8 @@ export function getSiteFragment( path: URLString ): SiteSlug | SiteId | false {
 		}
 	}
 
-	// Check last and second-to-last piece for site slug
-	for ( let i = 2; i > 0; i-- ) {
+	// Check last, second-to-last and third-to-last (in case of locale param)  piece for site slug
+	for ( let i = 3; i > 0; i-- ) {
```

Comment:
This patch is needed to make the
```
/checkout/jetpack/:siteSlug/:productSlug/:lang
```
route work correctly? You could avoid changing this function if the site param was called `:site` instead. The code that determines the site fragment looks like:
```js
siteFragment = context.params.site || getSiteFragment( context.path );
```
It first looks at the `:site` param, and only if it's not present, it does the unreliable heuristic in `getSiteFragment`.

Better to call the param `:site` and avoid the heuristic if possible.

Reply:
So it's actually this one: `/checkout/:product/:domainOrProduct/:lang` and in this case it's product. There are quite a lot of middleware functions (`redirectLoggedOut, siteSelection, redirectJetpackLegacyPlans, checkout, makeLayout, clientRender`) so I'm a bit worried I could broke something deeper inside? Or should we still do the refactoring here?

Reply:
Here the site slug is in `:domainOrProduct`, never in `:product`, is it? So looking at the second-to-last piece is enough, we don't need to look at the third.

What does the `/checkout/:product/:domainOrProduct` route display anyway? The `checkout` controller never reads the `:product` or `:domainOrProduct` params, it reads a set of different ones, none of which is present in the route.

Reply:
I'm very new to Calypso and maybe lacking some context. So, on `jetpack.cloud` side in `get-purchase-url-callback` I see that URL is build like: `/checkout/${ siteSlug }/${ productsString }` -> `/checkout/:product/:domainOrProduct` so `product = siteSlug`, and  `domainOrProduct = productsString`. It renders a checkout page for given site and product. So, in this case, I would probably need to rename 'product' into 'site', and then refactor all middleware functions like `getDomainOrProductFromContext` and make sure if any other registered pages rely on it.

Reply:
I think I undestand the checkout routes a bit better now. The routes affected by this PR, the ones that get a new `:lang` suffix, are:
- `/checkout/jetpack/:siteSlug/:productSlug`: here we can rename the params to `:site` and `:product` and change the places that read from `context.params`. There are just 2-3 of them. The benefits are that `siteSelection` will pick up the `:site` param automatically, without `getSiteFragment`, and that the `:product` param name is now more consistent with other checkout routes
- `/checkout/:domainOrProduct`: this is the only one where params are ambiguous: the route can be either `/checkout/example.blog`, showing checkout for the `example.blog` site, without adding any products, or `/checkout/plan_name`, selecting no site and adding `plan_name` to cart. Here the param can be renamed to `:site` and then we need to also change `getDomainOrProductFromContext` to read from `params.site` instead of `params.domainOrProduct`.
- `/checkout/:product/:domainOrProduct`: here the param naming seems completely wrong, but accidentally it all works. `:product` is always a site, and `siteSelection` will read it with `getSiteFragment`, no matter what the param name is. And the product slug will be read from `domainOrProduct` with `getDomainOrProductFromContext`. The route should be renamed to `/checkout/:site/:product`. Then `siteSelection` will read the site slug from `:site`, even if it's the third last (after `:lang` is appended), and `getDomainOrProductFromContext` will read product slug from `:product`.

What do you think? The proposed changes would clean up the param naming and would make the routing code more comprehensible.

Reply:
Thanks for the clarification. Will try to refactor this part.